### PR TITLE
Feat/slurm nodes when constraints enabled

### PR DESCRIPTION
**Pull request by tristan0x**

### CODE REVIEW

File: hpcbench/driver.py [hunk 1]
```diff
@@ -50,6 +50,7 @@
 )
 from . campaign import (
     from_file,
+    NodeSet,
```
Comment (ohm314) [inline hpcbench/driver.py]: I don't understand why you are importing this from campaign - shouldn't this come from `hpcbench.ext.ClusterShell`?

File: hpcbench/driver.py [hunk 2]
```diff
@@ -238,17 +232,23 @@ def nodes(self, tag):
 class CampaignDriver(Enumerator):
     """Abstract representation of an entire campaign"""
     def __init__(self, campaign_file=None, campaign_path=None,
+                 campaign=None,
                  node=None, output_dir=None, srun=None,
                  logger=None, expandcampvars=True):
         node = node or socket.gethostname()
-        if campaign_file and campaign_path:
-            raise Exception('Either campaign_file xor path can be specified')
+        if campaign_file and campaign_path and campaign:
```
Comment (ohm314) [inline hpcbench/driver.py]: if two of the three operands evaluate to True and one evaluates to False, the expression will still evaluate to False - so, I think this condition is not quite right

File: hpcbench/toolbox/contextlib_ext.py [hunk 3]
```diff
@@ -102,3 +102,35 @@ def mkdtemp(*args, **kwargs):
...
+    :param remove: Environment variables to remove.
+    :param update: Dictionary of environment variables
+    and values to add/update.
+    """
+    env = os.environ
+    update = update or {}
+    remove = remove or []
+
+    # List of environment variables being updated or removed.
+    stomped = (set(update.keys()) | set(remove)) & set(env.keys())
```
Comment (ohm314) [inline hpcbench/toolbox/contextlib_ext.py]: <nitpick>`set(update)` would also do </nitpick>

File: hpcbench/toolbox/contextlib_ext.py [hunk 4]
```diff
@@ -102,3 +102,35 @@ def mkdtemp(*args, **kwargs):
     finally:
         if remove:
             shutil.rmtree(path)
+
+
+@contextlib.contextmanager
+def modified_environ(*remove, **update):
```
Comment (ohm314) [inline hpcbench/toolbox/contextlib_ext.py]: Question: can you not use this cool utility now in driver for the `pre_execute` and `post_execute` ?

File: hpcbench/driver.py (hunk 1 above)
Comment (tristan0x) [reply hpcbench/driver.py]: Yes it is the same symbol. My point is that driver is using campaign. And campaign could decide to use another NodeSet implementation in the future, that wouldn't change logic in driver.
Does that make sense to you?

File: hpcbench/driver.py (hunk 2 above)
Comment (tristan0x) [reply hpcbench/driver.py]: You're right!

File: hpcbench/toolbox/contextlib_ext.py (hunk 3 above)
Comment (tristan0x) [reply hpcbench/toolbox/contextlib_ext.py]: you're right

File: hpcbench/toolbox/contextlib_ext.py (hunk 4 above)
Comment (tristan0x) [reply hpcbench/toolbox/contextlib_ext.py]: Yes. absolutely.

File: hpcbench/driver.py (hunk 1 above)
Comment (ohm314) [reply hpcbench/driver.py]: Ah, I see. Yes makes sense!

**Comment by tristan0x** [comment]
Hello @ohm314 I took your remarks in my last commits and also took the liberty to simplify CampaignDriver constructor's signature.
Feel free to merge, should you choose to accept it...

**Pull request merged**